src/ui/public/vis/request_handlers/courier.js

```diff
@@ -51,36 +40,35 @@ const CourierRequestHandlerProvider = function (Private, courier, timefilter) {
       });
 
       // Add the explicit passed timeRange as a filter to the requestSearchSource.
-      requestSearchSource.filter(() => {
+      timeFilterSearchSource.set('filter', () => {
         return timefilter.get(searchSource.index(), timeRange);
       });
 
-      removeSearchSourceParentTimefilter(requestSearchSource);
-
-      if (queryFilter && vis.editorMode) {
-        searchSource.set('filter', queryFilter.getFilters());
-        searchSource.set('query', appState.query);
+      // in editor mode we want to set filter and query on searchSource so they get
+      // stored with visualization, however on dashboard we set them on requestSearchSource
+      // so we don't override the stored ones
+      if (vis.editorMode) {
+        searchSource.set('filter', filters);
+        searchSource.set('query', query);
+      } else {
+        requestSearchSource.set('filter', filters);
+        requestSearchSource.set('query', query);
       }
 
-      const shouldQuery = () => {
+      const shouldQuery = async () => {
+        const currentRequest = await requestSearchSource.getSearchRequestBody();
         if (!searchSource.lastQuery || forceFetch) return true;
-        if (!_.isEqual(_.cloneDeep(searchSource.get('filter')), searchSource.lastQuery.filter)) return true;
-        if (!_.isEqual(_.cloneDeep(searchSource.get('query')), searchSource.lastQuery.query)) return true;
-        if (!_.isEqual(calculateObjectHash(vis.getAggConfig()), searchSource.lastQuery.aggs)) return true;
-        if (!_.isEqual(_.cloneDeep(timeRange), searchSource.lastQuery.timeRange)) return true;
+        if (searchSource.lastQuery !== calculateObjectHash(currentRequest)) return true;
 
         return false;
       };
 
-      return new Promise((resolve, reject) => {
-        if (shouldQuery()) {
+
+
+      return new Promise(async (resolve, reject) => {
```

Comment:
I strongly recommend sticking with either `.then/catch()` or `async/await` and not mixing them, and since this is using a lot of angular stuff and promises I think `.then/catch()` is a safer bet.